cpp/convert_json_to_marc.cc

```diff
@@ -0,0 +1,493 @@
+                                             subfield_prefix + GetScalarJSONStringValueWithoutQuotes(scalar_node_or_null));
+                }
+            }
+
+            for (const auto &subfield_code_and_fixed_subfield : field_descriptor.subfield_codes_to_fixed_subfields_)
+                new_field.appendSubfield(subfield_code_and_fixed_subfield.first, subfield_code_and_fixed_subfield.second);
+
+            record->insertField(new_field);
+            created_at_least_one_field = true;
+        } else { // All our data resides in JSON arrays.
+            for (unsigned json_array_index(0); json_array_index < array_length; ++json_array_index) {
+                std::string tag(field_descriptor.tag_);
+                if (json_array_index > 0 and not field_descriptor.overflow_tag_.empty())
+                    tag = field_descriptor.overflow_tag_;
+                MARC::Record::Field new_field(tag);
+
+                for (const auto &subfield_code_and_json_tag : field_descriptor.subfield_codes_to_json_tags_) {
+                    const auto node(object->getNode(subfield_code_and_json_tag.second));
+                    if (node == nullptr)
+                        continue;
+
+                    const std::string subfield_prefix(FindMapEntryForSubfieldCode(subfield_code_and_json_tag.first,
+                                                                                  field_descriptor.subfield_codes_to_prefixes_));
+                    const auto array_node(JSON::JSONNode::CastToArrayNodeOrDie("array_node", node));
+                    const auto scalar_node(array_node->getNode(json_array_index));
+                    new_field.appendSubfield(subfield_code_and_json_tag.first,
+                                             subfield_prefix + GetScalarJSONStringValueWithoutQuotes(scalar_node));
+                }
+
+                record->insertField(new_field);
+                created_at_least_one_field = true;
+            }
+        }
+    }
+
+final_processing:
+    if (field_descriptor.required_ and not created_at_least_one_field)
+        LOG_ERROR("required entry for \"" + field_descriptor.name_ + "\" not found!");
+}
+
+
+void GenerateSingleMARCRecordFromJSON(const std::shared_ptr<const JSON::ObjectNode> &object,
+                                      const JSONNodeToBibliographicLevelMapper &json_node_to_bibliographic_level_mapper,
+                                      const std::vector<FieldDescriptor> &field_descriptors, MARC::Writer * const marc_writer)
+{
+    std::string control_number;
+    const auto descriptor_for_field_001(std::find_if(field_descriptors.begin(), field_descriptors.end(),
+                                                     [](const FieldDescriptor &descriptor){ return descriptor.tag_ == "001"; }));
+    if (descriptor_for_field_001 != field_descriptors.end()) {
+        control_number = object->getOptionalStringValue(descriptor_for_field_001->json_tag_);
+        if (not control_number.empty())
```

Comment:
Shouldn't this be `if (control_number.empty())`?

Reply:
No, because if we have no data for a field we don't just want to shove the prefix into the field.  It's not the same as a default value.